### ISSUE TITLE
I was unable to test code for url.peg

### ISSUE THREAD

**Issue by machinshin** [jcoglan/canopy#46]
I tried going through the examples of javascript usage here: 
http://canopy.jcoglan.com/langs/javascript.html

but i get the following error: 
```sh
?> canopy url.peg --lang js
invalid options argument
TypeError: invalid options argument
    at optsArg (/usr/local/lib/node_modules/canopy/node_modules/mkdirp/lib/opts-arg.js:13:11)
    at mkdirp (/usr/local/lib/node_modules/canopy/node_modules/mkdirp/index.js:11:10)
    at write (/usr/local/lib/node_modules/canopy/bin/canopy:32:5)
    at Object.<anonymous> (/usr/local/lib/node_modules/canopy/bin/canopy:37:3)
    at Module._compile (internal/modules/cjs/loader.js:1139:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1159:10)
    at Module.load (internal/modules/cjs/loader.js:988:32)
    at Function.Module._load (internal/modules/cjs/loader.js:896:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
```

can anyone tell me what the issue here is?

**Comment by sumginazu** [comment]
@machinshin I just ran into the same issue. I suspect something in the npm package is out of date. I was able to work around this by building canopy from source, following the instructions in CONTRIBUTING.md. The executable created there works just fine.

**Comment by patb2** [comment]
Encountered same problem, demo examples do not work, throws invalid options argument error, and @sumginazu's workaround did not work either.  It's really a pity for such a reputable and widely used library.

**Comment by teichopsia** [comment]
fwiw - the problem is mkdirp was converted in 1.x to promises so the syntax invoked in canopy needs updating to

```diff
--- node_modules/canopy/bin/canopy      2020-11-23 01:17:52.459859400 -0800
+++ canopy-0.3.0-mkdirp-1.x-patch/canopy        2020-11-23 01:19:03.316873075 -0800
@@ -29,7 +29,7 @@
     var name = names[index];
     if (!name) return;
 
-    mkdirp(path.dirname(name), function(error) {
+    mkdirp(path.dirname(name)).then(function(){
       fs.writeFileSync(name, parser[name]);
       write(index + 1);
     });
```
so that's a quick patch if you feel like hacking the post npm installed result.

not sure if the author intends on releasing an interim patch to deal with it since 0.3.0 was 5 years back but that solved my use case to just get it running. ymmv. efforts to pin mkdirp to 0.5.1 fell down a npm hole where package{,-lock}.json and npm-shrinkwrap.json all stomp on one another and none of them allow regressing a specific dependency. if somebody figures out a solution to that so be it.

**Comment by nippur72** [comment]
stumbled on the same issue, since 0.4.0 is lagging can we just have a 0.3.1 with this fixed ?

**Comment by Bartvds** [comment]
I also had this error and got it working with the above patch. 

@jcoglan a fix release to keep current release installable would be nice.

**Comment by js850** [comment]
+1, I experienced this also when trying to run the tutorial.  The patch by teichopsia fixed it for me.

**Comment by jcoglan** [comment]
This issue is fixed in `master` so I'm going to close this. I haven't had time to work on this project for several months but I will publish new releases as soon as I have time.

**Comment by jcoglan** [comment]
This should now be fixed as part of the v0.4 release; see https://blog.jcoglan.com/2022/04/01/announcing-canopy-0-4/
